framework/src/modules/chain/synchronizer/block_synchronization_mechanism.js

```diff
@@ -42,47 +45,210 @@ class BlockSynchronizationMechanism {
+	 * of rounds (listSizeLimit)
+	 *
+	 * @param listSizeLimit - The size of the array to be computed
+	 * @param _currentRound
+	 * @return {Promise<*>}
+	 * @private
+	 */
+	async _computeBlockHeightsList(listSizeLimit, _currentRound) {
+		const blockHeights = [];
+		let numberOfBlocks = 0; // Keeps track of the number of block IDs to be included in a request
+		let currentRound = _currentRound;
+		let currentHeight = currentRound * this.constants.activeDelegates;
+
+		while (
+			numberOfBlocks < listSizeLimit &&
+			currentHeight > this.bft.finalizedHeight
+		) {
+			let height = currentRound * this.constants.activeDelegates;
+			if (height <= this.bft.finalizedHeight) {
+				// if the calculated height is smaller than finalized height we push finalized height
+				// on the array and currentRound -= 1 will stop the loop.
+				// The reason for this is stated in step 3.b of LIP-0014 under Block Synchronization Mechanism
+				height = this.bft.finalizedHeight;
 			}
-		} finally {
-			this.active = false;
+
+			blockHeights.push(height);
+			currentRound -= 1;
+			currentHeight = currentRound * this.constants.activeDelegates;
+			numberOfBlocks += 1;
+		}
+
+		return blockHeights;
+	}
+
+	/**
+	 * Requests the last common block in common with the targeted peer.
+	 * In order to do that, sends a set of network calls which include a set of block ids
+	 * corresponding to the first block of descendent consecutive rounds (starting from the last one)
+	 * @param peer - The peer to target.
+	 * @return {Promise<Object | undefined>}
+	 * @private
+	 */
+	async _requestLastCommonBlock(peer) {
+		const blocksPerRequestLimit = 10; // Maximum number of block IDs to be included in a single request
+		const requestLimit = 10; // Maximum number of requests to be made to the remote peer
+
+		let numberOfRequests = 0; // Keeps track of the number of requests made to the remote peer
+		let highestCommonBlock; // Holds the common block returned by the peer if found.
+		let currentRound = Math.floor(
+			this.blocks.lastBlock.height / this.constants.activeDelegates,
+		); // Holds the current round number
+		let currentHeight = currentRound * this.constants.activeDelegates;
+
+		while (
+			!highestCommonBlock &&
+			numberOfRequests < requestLimit &&
+			currentHeight > this.bft.finalizedHeight
+		) {
+			const blockIds = (await this.storage.entities.Block.get(
+				{
+					height_in: this._computeBlockHeightsList(
+						blocksPerRequestLimit,
+						currentRound,
+					),
+				},
+				{
+					sort: 'height:asc',
+				},
+			)).map(block => block.id);
+
+			// Request the highest common block with the previously computed list
+			// to the given peer
+			const { data } = await this.channel.invoke('network:requestFromPeer', {
+				procedure: 'getHighestCommonBlock',
+				peerId: peer.id,
+				data: blockIds,
+			});
+
+			highestCommonBlock = data; // If no common block, data is undefined.
```

Comment:
no need to reassign here.
maybe like
`const { data: highestCommonBlock } = ...`?

Reply:
It is reassigned in order to check if the variable holds any value. If it does, the outer while loop stops.